rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -71,12 +73,22 @@ class DebruijnInterpreter[M[_], F[_]](
     */
   private def produce(
       chan: Par,
-      data: Seq[Par],
+      data: ListParWithRandom,
       persistent: Boolean,
-      rand: Blake2b512Random,
       sequenceNumber: Int
-  ): M[Unit] =
-    tuplespaceAlg.produce(chan, ListParWithRandom(data, rand), persistent, sequenceNumber)
+  ): M[Unit] = {
+    def go: Option[(TaggedContinuation, Seq[ListParWithRandom], Int)] => M[Unit] = {
+      case Some((continuation, dataList, updatedSequenceNumber)) =>
+        if (persistent)
+          List(
+            dispatcher.dispatch(continuation, dataList, updatedSequenceNumber),
+            produce(chan, data, persistent, sequenceNumber)
+          ).parSequence_
+        else dispatcher.dispatch(continuation, dataList, updatedSequenceNumber)
+      case None => s.unit
+    }
+    pureRSpace.produce(chan, data, persist = persistent, sequenceNumber) >>= (go(_))
```

Comment:
It's nice to have this logic in the only place it is used :+1:

Please restore the `go` method (vs lambda) though :)

Reply:
If you're after concision, extracting a type alias/case class for the tuplespace return type could be a good way :)